platform/cms/accelerated_publishing/rfc/broken-link-checking.md

```diff
@@ -0,0 +1,126 @@
+- This sort of checker would create additional traffic to the site. This should not present a problem to our current S3-as-web server architecture, but we should validate that the reverse proxy will not have trouble with the additional traffic, and we should ensure this traffic is ignored by analytics systems.
+
+### Build time Proposal
+
+We could check links on pages as they are built. However, these checks would need to check against the live production site, which requires making a request to a URL and verifying the response. This is _significantly_ slower than the current Content Build system.
+
+It is only recommended that we do this when we move to an incremental publishing scenario, and also that the link checks happen independently of deployment. In other words, link checks would be triggered by publishing of pages, but they would happen after deployment, rather than before.
+
+Advantages:
+
+- Errors would be caught at build time and rectified sooner.
+
+Challenges:
+
+- Publishing new groups of pages that link to each other could present challenges, as they are not on production yet and cannot be checked there.
+- Broken link reports would come in quick bursts, which could present challenges to CMS User Support, who are currently responsible for triaging broken link issues.
+
+### Combined proposal
+
+These two proposals could both be put into use and compliment each other, each used appropriately.
+
+For example, the situation which presents the greatest liklihood for larger scale broken link issues are code changes. Therefore, any deployment of code changes to Next Build should have a full broken link scan as part their staging process and after deploy, one that checks both Next Build and Content Build content.
+
+Build-time checking is more appropriate for small content changes througout the day. We would still recommend a full scan nightly, to catch anything that may have not been caught by small-scale scans.
+
+## Risks
+
+_What are the risks of the proposed changes?_
+
+Some of these risks have been discussed above.
+
+1. The current system (build everything and check everything) is inherently a problem for speedy publishing, but it _does_ check every internal link every publishing run. What we are proposing would actually capture more broken links than the current system, but there is potential for a broken link to be live on VA.gov for longer before it is reported.
+2. The current system has existed in its current form for some time and over time has reached something like stability. We would be building a new system, and inevitably there will be issues that will need to be resolved as the new system comes into use.
+3. Depending on architecture choices, we risk reintroducing slowdown into the publishing process. We want to avoid this as much as possible while ensuring the integrity of linked content on VA.gov.
+
+## Alternatives
+
+_What other alternatives solutions were considered, why weren't they chosen?_
```

Comment:
[A similar proposal was pitched by this design doc](https://github.com/department-of-veterans-affairs/va.gov-team/pull/14092) several years ago.
It accounted for notifications to editors.

This proposal does not seem to account for notification of anyone.  Would this tie into the same notification system we currently have (Slack notifications for CMS Helpdesk)?

Reply:
I will add something to speak to this question. I think addressing notifications is outside the scope of this proposal, and so we would default to replicating the current notification system, I.E. Slack notifications that are monitored by CMS User Support. 

User Support folks have said that the current system is not at all ideal for them, because it's a lot of repetitive notification of the same broken links with every content build, and each one needs to be scanned for new issues mixed in with the existing ones, which presents a significant cognitive load. 

Editor notifications would be a dream, if we can get them to respond to them. It would need to be emails; or at least, the research for outdated content notifications indicated that editors vastly prefer email notification to anything else. 

In either case, I think whatever broken link checks we come up with should produce machine-formatted results that can be ingested by other systems. I think User Support could greatly benefit from a dashboard-type application that helps them keep track of broken links and allows them to clearly see both new broken links as well as their age, i.e. how long it takes to resolve them. 

The same data could also be used to drive editor notifications, though I think a lot of user research would need to be done there to know what the best way to notify editors is.

Reply:
I will add something to speak to this question. I think addressing notifications is outside the scope of this proposal, and so we would default to replicating the current notification system, I.E. Slack notifications that are monitored by CMS User Support. 

User Support folks have said that the current system is not at all ideal for them, because it's a lot of repetitive notification of the same broken links with every content build, and each one needs to be scanned for new issues mixed in with the existing ones, which presents a significant cognitive load. 

Editor notifications would be a dream, if we can get them to respond to them. It would need to be emails; or at least, the research for outdated content notifications indicated that editors vastly prefer email notification to anything else. 

In either case, I think whatever broken link checks we come up with should produce machine-formatted results that can be ingested by other systems. I think User Support could greatly benefit from a dashboard-type application that helps them keep track of broken links and allows them to clearly see both new broken links as well as their age, i.e. how long it takes to resolve them. 

The same data could also be used to drive editor notifications, though I think a lot of user research would need to be done there to know what the best way to notify editors is.

Reply:
You already laid the groundwork for the scan to be machine readable.  Drupal could easily ingest that (it was supposed to be the next step) and put notifications in a dashboard.  I do not recommend email for this.  You'd send out 5 emails to editors in one section.  one person would fix it and the others would be confused by their copy of the email that cried wolf.

One email a month for outdated content does not create much cognitive overload, but 20 a day about broken links certainly would.

Reply:
Regardless, notifications aren't part of this proposal. That'll be its own project at some point.